src/mame/video/tmap038.cpp

```diff
@@ -0,0 +1,266 @@
+
+        2.w         f--- ---- ---- ----     0 = Layer Flip Y
+                    -e-- ---- ---- ----     Activate Row-select
+                    --d- ---- ---- ----     0 = 8x8 tiles, 1 = 16x16 tiles
+                    ---c ba9- ---- ----
+                    ---- ---8 7654 3210     Scroll Y
+
+        4.w         fedc ba98 765- ----
+                    ---- ---- ---4 ----     Layer Disable
+                    ---- ---- ---- 3210     Varies*
+
+		*color bank for mcatadv or Layer-Layer priority for cave
+
+There are more!
+
+***************************************************************************
+
+TODO:
+	de-fragmentation and merge drawing behavior into tmap038.cpp
+
+***************************************************************************
+
+*/
+
+#include "emu.h"
+#include "tmap038.h"
+
+void tilemap038_device::vram_map(address_map &map)
+{
+	map(0x0000, 0x0fff).rw(FUNC(tilemap038_device::vram_16x16_r), FUNC(tilemap038_device::vram_16x16_w)).share("vram_16x16");
+	map(0x1000, 0x17ff).rw(FUNC(tilemap038_device::lineram_r), FUNC(tilemap038_device::lineram_w)).share("lineram");
+	map(0x1800, 0x3fff).ram().share("scratchpad"); // scratchpad?
+	map(0x4000, 0x7fff).rw(FUNC(tilemap038_device::vram_8x8_r), FUNC(tilemap038_device::vram_8x8_w)).share("vram_8x8");
+}
+
+void tilemap038_device::vram_writeonly_map(address_map &map)
+{
+	map(0x0000, 0x0fff).w(FUNC(tilemap038_device::vram_16x16_w)).share("vram_16x16");
+	map(0x1000, 0x17ff).w(FUNC(tilemap038_device::lineram_w)).share("lineram");
+	map(0x1800, 0x3fff).writeonly().share("scratchpad"); // scratchpad?
+	map(0x4000, 0x7fff).w(FUNC(tilemap038_device::vram_8x8_w)).share("vram_8x8");
+}
+
+void tilemap038_device::vram_16x16_map(address_map &map)
+{
+	map(0x0000, 0x0fff).rw(FUNC(tilemap038_device::vram_16x16_r), FUNC(tilemap038_device::vram_16x16_w)).share("vram_16x16");
+	map(0x1000, 0x17ff).rw(FUNC(tilemap038_device::lineram_r), FUNC(tilemap038_device::lineram_w)).share("lineram");
+	map(0x1800, 0x3fff).ram().share("scratchpad"); // scratchpad?
+}
+
+void tilemap038_device::vram_16x16_writeonly_map(address_map &map)
+{
+	map(0x0000, 0x0fff).w(FUNC(tilemap038_device::vram_16x16_w)).share("vram_16x16");
+	map(0x1000, 0x17ff).w(FUNC(tilemap038_device::lineram_w)).share("lineram");
+	map(0x1800, 0x3fff).writeonly().share("scratchpad"); // scratchpad?
+}
+
+/*  Some games, that only ever use the 8x8 tiles and no line scroll,
+    use mirror ram. For example in donpachi, writes to 400000-403fff
+    and 408000-407fff both go to the 8x8 tilemap ram. Use this function
+    in this cases. Note that the get_tile_info function looks in the
+    4000-7fff range for tiles, so we have to write the data there. */
+void tilemap038_device::vram_8x8_map(address_map &map)
+{
+	map(0x0000, 0x3fff).mirror(0x4000).rw(FUNC(tilemap038_device::vram_8x8_r), FUNC(tilemap038_device::vram_8x8_w)).share("vram_8x8");
+}
+
+DEFINE_DEVICE_TYPE(TMAP038, tilemap038_device, "tmap038", "038 Tilemap generator")
```

Comment:
Device name is too generic and missing manufacturer, also is it actually from the PCB chip markings?

Reply:
The device itself has very few markings and doesn't say who manufactured it. (Given the dates of the earliest games that use it, I suspect it might not actually have been by Cave.)

Reply:
Manufactured by NEC, but I can't find a chip designer. (Might be Atlus?)